drivers/SmartThings/matter-switch/src/switch_utils/utils.lua

```diff
@@ -297,50 +318,86 @@ function utils.detect_bridge(device)
+  return true
 end
 
-function utils.report_power_consumption_to_st_energy(device, latest_total_imported_energy_wh)
+function utils.report_power_consumption_to_st_energy(device, endpoint_id, total_imported_energy_wh)
   local current_time = os.time()
   local last_time = device:get_field(fields.LAST_IMPORTED_REPORT_TIMESTAMP) or 0
 
   -- Ensure that the previous report was sent at least 15 minutes ago
   if fields.MINIMUM_ST_ENERGY_REPORT_INTERVAL >= (current_time - last_time) then
     return
   end
-
   device:set_field(fields.LAST_IMPORTED_REPORT_TIMESTAMP, current_time, { persist = true })
 
-  -- Calculate the energy delta between reports
-  local energy_delta_wh = 0.0
-  local previous_imported_report = device:get_latest_state("main", capabilities.powerConsumptionReport.ID,
-    capabilities.powerConsumptionReport.powerConsumption.NAME)
-  if previous_imported_report and previous_imported_report.energy then
-    energy_delta_wh = math.max(latest_total_imported_energy_wh - previous_imported_report.energy, 0.0)
-  end
-
+  local state_device = utils.find_child(device, endpoint_id) or device
+  local previous_imported_report = state_device:get_latest_state("main", capabilities.powerConsumptionReport.ID,
+    capabilities.powerConsumptionReport.powerConsumption.NAME, { energy = total_imported_energy_wh }) -- default value if nil
+  -- Report the energy consumed during the time interval. The unit of these values should be 'Wh'
   local epoch_to_iso8601 = function(time) return os.date("!%Y-%m-%dT%H:%M:%SZ", time) end -- Return an ISO-8061 timestamp from UTC
+  device:emit_event_for_endpoint(endpoint_id, capabilities.powerConsumptionReport.powerConsumption({
+    start = epoch_to_iso8601(last_time),
+    ["end"] = epoch_to_iso8601(current_time - 1),
+    deltaEnergy = total_imported_energy_wh - previous_imported_report.energy,
+    energy = total_imported_energy_wh
+  }))
+end
 
-  -- Report the energy consumed during the time interval. The unit of these values should be 'Wh'
-  if not device:get_field(fields.ENERGY_MANAGEMENT_ENDPOINT) then
-    device:emit_event(capabilities.powerConsumptionReport.powerConsumption({
-      start = epoch_to_iso8601(last_time),
-      ["end"] = epoch_to_iso8601(current_time - 1),
-      deltaEnergy = energy_delta_wh,
-      energy = latest_total_imported_energy_wh
-    }))
+function utils.set_fields_for_electrical_sensor_endpoint(device, electrical_ep_info, associated_endpoint_ids)
+  local tags = ""
+  if utils.find_cluster_for_ep(electrical_ep_info, clusters.ElectricalPowerMeasurement.ID) then tags = tags.."-power" end
+  if utils.find_cluster_for_ep(electrical_ep_info, clusters.ElectricalEnergyMeasurement.ID) then tags = tags.."-energy-powerConsumption" end
+  if associated_endpoint_ids ~= {} then
+    table.sort(associated_endpoint_ids)
+    local primary_associated_ep_id = type(associated_endpoint_ids[1]) == "table" and associated_endpoint_ids[1].value or associated_endpoint_ids[1]
```

Comment:
Can we define an expectation for the value of the `associated_endpoint_ids` parameter so that this check is unnecessary? The caller can ensure that the data is in the correct format, which essentially means that we ensure that since this is an internal helper function. It might be somewhat cleaner, and would clearly define the expected input.